File: app/src/main/java/com/tr4n/demolistadapter/java/PhotoAdapter.java
package com.tr4n.demolistadapter.java;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.tr4n.demolistadapter.R;

import java.util.ArrayList;
import java.util.List;

public class PhotoAdapter extends RecyclerView.Adapter<PhotoViewHolder> {

    private List<Photo> photos = new ArrayList<>();
    private OnItemClickListener listener;

    @NonNull
    @Override
    public PhotoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_photo, parent, false);
        return new PhotoViewHolder(itemView, listener);
    }

    @Override
    public void onBindViewHolder(@NonNull PhotoViewHolder holder, int position) {
        holder.bindData(photos.get(position));
    }

    @Override
    public int getItemCount() {
        return photos != null ? photos.size() : 0;
    }

    public void setItemClickListener(OnItemClickListener listener) {
        this.listener = listener;
    }

    public void updateData(List<Photo> newPhotos) {
        photos.clear();
        photos.addAll(newPhotos);
        notifyDataSetChanged();
    }

    public void addData(List<Photo> morePhotos) {
        int oldSize = getItemCount();
        photos.addAll(morePhotos);
        int newSize = getItemCount();
        notifyItemRangeChanged(oldSize, newSize);
    }
}

File: app/src/main/java/com/tr4n/demolistadapter/kotlin/PhotoAdapter.kt
package com.tr4n.demolistadapter.kotlin

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.tr4n.demolistadapter.R

class PhotoAdapter : RecyclerView.Adapter<PhotoViewHolder>() {

    var onItemClick: (Photo, Int) -> Unit = { _, _ -> }
    private val photos = mutableListOf<Photo>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PhotoViewHolder {
        val itemView =
            LayoutInflater.from(parent.context).inflate(R.layout.item_photo, parent, false)
        return PhotoViewHolder(itemView, onItemClick)
    }

    override fun getItemCount(): Int = photos.size

    override fun onBindViewHolder(holder: PhotoViewHolder, position: Int) {
        holder.bindData(photos[position])
    }

    fun updateData(newPhotos: List<Photo>) {
        photos.clear()
        photos.addAll(newPhotos)
        notifyDataSetChanged()
    }

    fun addData(morePhotos: List<Photo>) {
        val oldSize = itemCount
        photos.addAll(morePhotos)
        val newSize = itemCount
        notifyItemRangeChanged(oldSize, newSize)
    }
}